### DISCUSSION TITLE
How I can change data in prisma.schema, without losing data after migration?

### DISCUSSION THREAD

**Discussion by mxcdh** [prisma/prisma#13160] [Q&A]
When I changed prisma.schema and run:
```
npx prisma migrate dev --name init
```

My Postgres lost all data. Can I run another command which doesn't remove my data? What is best for me in this situation?

**Comment by nurul3101** [comment]
Hey @mxcdh 👋 

For quick prototyping, you could use the `npx prisma db push` command. Here's a reference for using [db push](https://www.prisma.io/docs/guides/database/prototyping-schema-db-push)
Here's a guide on when you should choose [migrate or db push](https://www.prisma.io/docs/concepts/components/prisma-migrate/db-push#choosing-db-push-or-prisma-migrate).

**Reply by moshibbin** [reply]
very nice

**Comment by janpio** [comment]
> My Postgres lost all data.

Why did it loose all data? Did you execute a reset? What was in that migration that you created and then executed?

**Comment by muhammad-Osman** [comment]
i have added one field in schema.prisma.
when i am trying to alter table on production it is giving me warning you will loss all data.

can you please suggest better solution

**Reply by janpio** [reply]
Was your migration history in sync with your Prisma schema before you added the one field?

**Reply by muhammad-Osman** [reply]
yes my migration history in sync with my Prisma schema

**Reply by janpio** [reply]
Then this is worth a bug report issue. If before adding the field you could run `migrate dev` with not changes detected, and then after adding a field you get a reset warning, that is most probably unintentional.

**Comment by gregory-krieger** [comment]
Just a quick update on this :
If you're trying to apply database changes which are not in compatible with data which is already stored, if you use :

`npx prisma db push`

It will warn that if you continue, you'll loose all the data.

For example, if you're trying to add a field in a table which is not optional, and you're do not provide a default value for the field, Prisma + MySql will not know what to do with the existing records of the table which will have no value in a field which is not optionnal (NOT NULL) and will therefore throw the warning.

Therefore, what you shoud :

**Option 1:**  Just add a default value.


**Option 2:**  Mark the field you're adding to a table with existing records as optional, knowing that you can then: 

1. Update the schema, marking the field as optional
2. Add the data to the added field manually
3. Update the schema, removing the optional attribute.

Fields, which do not have to be optional, should not be marked as optional, so that you can get clean types while coding.

**Comment by bamidele2161** [comment]
Hello @gregory-krieger,

Thank you for your suggestion.

In a scenario where two different databases are being used, each with its schema (Schema A and Schema B), and there's a need to make changes to fields of a table in Schema B, the challenge arises when running the migration command:

npx prisma migrate dev --schema prisma/schema2.prisma --name edit-form-datable --create-only

Even though the --create-only flag is used to generate migration files without applying the changes to the database immediately, it still prompts with a warning: "Do you want to continue? All data will be lost."

In this case, what steps can be taken to prevent potential data loss when making changes to Schema B?

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
